File: LeetCode/cpp/31.cpp
class Solution {
public:
    void nextPermutation(vector<int> &nums) {
        int size = nums.size(), n = nums.size() - 2;
        while (n >= 0) {
            if (nums[n] < nums[n + 1])
                break;
            --n;
        }
        if (n >= 0) {
            int sub = 0, upper = INT_MAX;
            for (int i = n + 1; i < size; ++i) {
                if (nums[i] > nums[n] && upper > nums[i]) {
                    sub = i;
                    upper = nums[i];
                }
            }
            swap(nums[sub], nums[n]);
        }
        std::sort(nums.begin() + n + 1, nums.end());
    }
};

File: LeetCode/python3/31.py
import sys


class Solution:
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        m, size = len(nums) - 2, len(nums)
        while m >= 0:
            if nums[m] < nums[m + 1]:
                break
            m -= 1
        if(m >= 0):
            upper, index = sys.maxsize, 0
            for i in range(m + 1, size):
                if nums[i] > nums[m] and upper > nums[i]:
                    upper = nums[i]
                    index = i
            nums[m], nums[index] = nums[index], nums[m]
        nums[m + 1: size] = sorted(nums[m + 1: size])
